### PR TITLE
[WIP] Moving API Manager to Knative/Istio

### PR DESCRIPTION
**Pull request by nickaashoek**
Goals:
- Get rid of async
- Use istio to manager our APIs

TODO:
- Authentication/HTTPs
- Fully support all features of istio virtual services
- Install using KO

### CODE REVIEW

File: pkg/api-manager/istiohandler.go [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+	}
+	return endpoint.NewGetApisOK().WithPayload(results)
+}
+
+func (cl *IstioHandlers) UpdateAPI(params endpoint.UpdateAPIParams, principal interface{}) middleware.Responder {
+	span, _ := trace.Trace(params.HTTPRequest.Context(), "")
+	defer span.Finish()
+
+	log.Infof("Trying to update: %+v", params)
+	return cl.AddAPI(endpoint.AddAPIParams{
```
Comment (berndtj) [inline pkg/api-manager/istiohandler.go]: Does update work?  Do we need to delete the endpoint first?

File: pkg/api-manager/istio/istio.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	crdClient, err := crd.NewClient("", "", []model.ProtoSchema{model.VirtualService}, "")
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		istioClient: crdClient,
+	}, nil
+}
+
+type Client struct {
```
Comment (imikushin) [inline pkg/api-manager/istio/istio.go]: You don't really need another layer of istio client abstraction: there's already an API client that you're using here in the `istioClient` field.

File: pkg/api-manager/istiohandler.go [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+}
+
+func (cl *IstioHandlers) AddAPI(params endpoint.AddAPIParams, principal interface{}) middleware.Responder {
+	span, ctx := trace.Trace(params.HTTPRequest.Context(), "")
+	defer span.Finish()
+
+	log.Infof("Trying to add istio api: %+v", params)
+
+	api := apiModelOntoIstioEntity(params.Body)
+	virtualServiceSpec, err := yaml.Marshal(api)
```
Comment (imikushin) [inline pkg/api-manager/istiohandler.go]: Separate Istio data model and YAML marshaling/unmarshaling seems redundant: why not just use istioClient directly?

File: pkg/api-manager/types.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
+package apimanager
+
+type VirtualService struct {
```
Comment (imikushin) [inline pkg/api-manager/types.go]: Looks like these are redundant if you use istioClient directly

**Comment by imikushin** [comment]
Looks much simpler now!

**Comment by berndtj** [comment]
The code has been merged in

**Pull request closed without merging**